### ISSUE TITLE
esp32_rom.elf这个文件需要自己编么？在哪里能弄到呢

### ISSUE THREAD

**Issue by zhaoyanbai** [espressif/qemu#1]
https://github.com/espressif/qemu/wiki
请问这个文章里提到的`esp32_rom.elf`在哪里？

**Comment by cbontoz** [comment]
I downloaded the esp32_rom.elf from ESP32 Core Dump documentation section ([Download Link](https://dl.espressif.com/dl/esp32_rom.elf)). But I receive the following error:

```
Adding SPI flash device
==946==WARNING: ASan doesn't fully support makecontext/swapcontext functions and may produce false positives in some cases!
==946==WARNING: ASan is ignoring requested __asan_handle_no_return: stack top: 0x7ffc7530e000; bottom 0x7f39d95fe000; size: 0x00c29bd10000 (835837820928)
False positive error reports may follow
For details see http://code.google.com/p/address-sanitizer/issues/detail?id=189
rom: requested regions overlap (rom phdr #22: buildqemu/esp32_rom.elf. free=0x000000004000d670, addr=0x000000003ffae020)
qemu-system-xtensa: rom check and register reset failed
```

Please let me know if you have made any progress on it.

**Comment by zhaoyanbai** [comment]
i have the same situation as you.

```
Adding SPI flash device
==61588==WARNING: ASan is ignoring requested __asan_handle_no_return: stack top: 0x7000067ac000; bottom 0x000110ef9000; size: 0x6ffef58b3000 (123140831916032)
False positive error reports may follow
For details see https://github.com/google/sanitizers/issues/189
rom: requested regions overlap (rom phdr #22: esp32_rom.elf. free=0x000000004000d670, addr=0x000000003ffae020)
```

**Comment by igrr** [comment]
Thanks for the report, the elf file on dl.espressif.com looks broken indeed. The section content doesn't match the real ROM code either. Will check.

**Comment by cbontoz** [comment]
Hi igrr, zhaoyanbai,

Have you had any progress on this? 

I generated a new ROM file using binaries read from my ESP32 EVK board, based on the [esp32-elf-rom](https://github.com/gschorcht/esp32-elf-rom) project. I do not receive the `rom check and register reset failed` now, but the gdb is stack in 0x400003c0 and the QEMU prints the following message:

> qemu-system-xtensa: GLib: g_mapped_file_unref: assertion 'file != NULL' failed

Any ideas what this might be? It seems like the `elf` file I generated is not recognized as valid `elf` by qemu. 

Thank you,
Chris

**Comment by Ebiroll** [comment]
None of the above rom.elf files are complete,
To make qemu run with a rom dump, you can do this, 
esptool.py --chip esp32 -b 921600 -p /dev/ttyUSB0 dump_mem 0x40000000 0x000C2000 rom.bin
This will also dump some areas after the rom

Then patch the function in hw/extensa/esp32.c
#include "qemu-common.h"

static void esp32_soc_init(Object *obj) {
... // Add at the end 
    const char *rom_filename = "rom.bin";

    rom_filename = qemu_find_file(QEMU_FILE_TYPE_BIOS, rom_filename);
    if (!rom_filename ||
        load_image_targphys(rom_filename, 0x40000000, 794624) < 0) {
        error_report("unable to load ROM image '%s'\n", rom_filename);
        exit(EXIT_FAILURE);
    }
}
 Start with,
xtensa-softmmu/qemu-system-xtensa -nographic  -machine esp32   -drive file=flash_image.bin,if=mtd,format=raw   -nographic

Also look here for more info,
https://github.com/Ebiroll/qemu_esp32

**Comment by Ebiroll** [comment]
Here is the diff of the patch https://github.com/Ebiroll/qemu-xtensa-esp32/commit/55d4a8955207c368608c5a873fb92012a012bcfe
I tried using a smaller dump but could not get it to work.
For those interested in investigating, 
    (gdb) awatch *0x3FF12000
    (gdb) awatch *0x3FF10000
 Sets a breakpoints where the boot-rom modifies the  Flash MMU table

**Comment by antfobe** [comment]
Yeah, while running ebirolls' patch above a memory leak warning pops to stdout:
![image](https://user-images.githubusercontent.com/13102173/68817343-283dfe80-0660-11ea-9ce0-caedc24016b7.png)

Im following the steps at "Enabling Ethernet support" [https://github.com/espressif/qemu/wiki](https://github.com/espressif/qemu/wiki) - set `CONFIG_EXAMPLE_CONNECT_ETHERNET` and `CONFIG_EXAMPLE_USE_OPENETH`
The example is `esp-idf/examples/protocols/http_server/simple`

**Comment by antfobe** [comment]
And using esptool to flash qemu is also not working:
![image](https://user-images.githubusercontent.com/13102173/68817973-2ffea280-0662-11ea-93d8-df480d31d8c4.png)

**Comment by Ebiroll** [comment]
@antfobe Do *NOT* use the -kernel parameter to start qemu.
Use the make-flash-img.sh to create the flash image. https://github.com/espressif/qemu/wiki
Then you can start qemu with
    xtensa-softmmu/qemu-system-xtensa -nographic -M esp32 -drive file=flash_image.bin,if=mtd,format=raw -nographic

**Comment by Ebiroll** [comment]
Make sure to configure
Top) → Component config → Ethernet
                                            Espressif IoT Development Framework Configuration
[ ] Support ESP32 internal EMAC controller  ----
[ ] Support SPI to Ethernet Module  ----
[*] Support OpenCores Ethernet MAC (for use with QEMU)  --->

And
(Top) → Example Connection Configuration
                                             Espressif IoT Development Framework Configuration
    Connect using (Ethernet)  --->
The make-flash-img.sh is broken on my system, but can be patched
When I run I got, 
Guru Meditation Error: Core  0 panic'ed (LoadStorePIFAddrError)
This because the wrong driver was initilzed.
After patching esp-idf/examples/common_components/protocol_examples_common/connect.c
I was able to run simple as,
    qemu-system-xtensa -nographic -M esp32 -drive file=flash_image.bin,if=mtd,format=raw -nographic -nic user,model=open_eth,hostfwd=tcp::10080-:80  -s  
Then connect browser to 127.0.0.1:10080

**Comment by igrr** [comment]
Thanks for the patience everyone, the esp-develop branch has been updated to include the ROM code (https://github.com/espressif/qemu/commit/86ba82e90fff3c814c4ab4db0bd28e34cd51c759). The `-kernel` argument is not required now. It can still be used to specify a different ROM code ELF file, if necessary.

The `make-flash-img.sh` script has also been updated, previously it failed with versions of `dd` which didn't support hexadecimal sizes.

I've also added a section in the Wiki about disabling hardware crypto support in menuconfig.
